### PR TITLE
exp/ingest: Ingest Session

### LINKED ISSUE #1310: Ingestion system and session
When the required components of the new ingestion system are ready:
- [x] `ledgerbackend` package and adapters https://github.com/stellar/go/issues/1285
- [x] `historyarchive` package, read/write interfaces https://github.com/stellar/go/issues/1134
- [x] `ingest/pipeline` package https://github.com/stellar/go/issues/961

we need to connect it all together using `ingest.System` and `ingest.Session` as described in https://github.com/stellar/go/blob/master/services/horizon/internal/docs/plans/new_horizon_ingest.md#tying-it-all-together.

### PR DESCRIPTION
**Pull request by bartekn**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ x This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [x] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [x] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [x] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### Summary

This PR adds two `Session` implementations (`LiveSession` and `SingleLedgerSession`) and a simple `horizon-demo` tool.

### Goal and scope

The goal of this PR is to implement the last missing component of `exp/ingest` that connects all the existing packages together: `Session`. `Session` is connecting to history archives and/or ledger backend and passes data to one or two pipelines (state pipeline and ledger pipeline).

`Session` supports one of the use cases developers can interact with Stellar ledger. For example: `LiveSession` initializes the state and then follows the new ledgers and processes transactions (it's running indefinitely). On the contrary `SingleLedgerSession` processes the state of a single ledger and terminates. More sessions will be added in a future (ex. `RangeSession` that processes data between two ledgers).

It also contains a simple demo app called `horizon-demo` (`go run ./exp/tools/horizon-demo`) that's using `LiveSession` internally. `horizon-demo` is reading data from history archives and ledger backend and 1) updates accounts for signers, 2) inserts transactions to a database and 3) updates in-memory orderbook graph.

Close #1310.

### Summary of changes

* Added `Session` implementations and `horizon-demo` app.
* Added `Reset()` method to pipeline processors. It's required to reset internal state of processors when a pipeline is used again (ex. when new ledger is closed). I'm still not sure if this is the best approach - read in the next section.
* Updated `support/pipeline.Pipeline` to be reusable (previously it was a single use only object) and added `Shutdown()` method.
* Added a few common processors to `ingest/processors` (`CSVPrinter` is an interesting example because it's possible to use it in both: state and ledger pipelines - it implements both interfaces).
* Updated `support/pipeline` with pre- and post-processing hooks that are useful for opening and committing database transactions but also applying changes to other structures (like orderbook graph). Maybe it's a solution for #1459?

### Known limitations & issues

* I'm still not sure if the current approach to resetting pipeline is good. It's elegant but I'm afraid that developers may implement `Reset()` incorrectly when creating a new processor. The alternative is to create something like a `PipelineFactory` that would create a completely new pipeline with default `struct` values.
* `adapters` package will likely be removed. It doesn't really give any value and acts as an unnecessary wrapper around readers.
* No docs and tests. Will be improved in another PR.

### What shouldn't be reviewed

Please ignore changes to `adapters` package. It's clear to me right now that we will remove and refactor it (#1405).

### CODE REVIEW

File: exp/ingest/io/ledger_read_closer.go [hunk 1]
```diff
@@ -92,7 +91,7 @@ func (dblrc *DBLedgerReadCloser) init() error {
 		return errors.Wrap(err, "error reading ledger from backend")
 	}
 	if !exists {
-		return errors.Wrap(err, "ledger was not found")
+		return ErrNotFound
 	}
```
Comment (bartekn) [inline exp/ingest/io/ledger_read_closer.go]: Interesting note: it was actually always returning `nil` before that change. `err` in line 95 was always `nil` and when `nil` is passed to `errors.Wrap` it returns `nil` as well. I wonder if this is checked by `staticcheck` because I found very similar instance of this bug earlier this week (https://github.com/stellar/go/pull/1443#discussion_r296744859). If not, would be great to add a rule that catches this.

File: exp/ingest/io/ledger_read_closer.go [hunk 2]
```diff
@@ -33,11 +24,19 @@ type DBLedgerReadCloser struct {
...
-func MakeLedgerReadCloser(sequence uint32, backend ledgerbackend.LedgerBackend) *DBLedgerReadCloser {
-	return &DBLedgerReadCloser{
+func MakeLedgerReadCloser(sequence uint32, backend ledgerbackend.LedgerBackend) (*DBLedgerReadCloser, error) {
+	reader := &DBLedgerReadCloser{
 		sequence: sequence,
 		backend:  backend,
 	}
+
+	var err error
+	reader.initOnce.Do(func() { err = reader.init() })
```
Comment (tamirms) [inline exp/ingest/io/ledger_read_closer.go]: related to https://github.com/stellar/go/issues/1433 . because `reader` is created a few lines above. it seems strange `reader` initialize it using `initOnce.Do()` . fixing this might be outside the scope of this PR though

File: exp/ingest/live_session.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	"fmt"
+	"time"
+
+	"github.com/stellar/go/exp/ingest/adapters"
+	"github.com/stellar/go/exp/ingest/io"
+	"github.com/stellar/go/exp/ingest/pipeline"
+	"github.com/stellar/go/support/errors"
+)
+
+var _ Session = &LiveSession{}
```
Comment (tamirms) [inline exp/ingest/live_session.go]: why is this necessary?

File: exp/ingest/live_session.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	s.ensureRunOnce()
+
+	historyAdapter := adapters.MakeHistoryArchiveAdapter(s.Archive)
+
+	s.currentLedger, err = historyAdapter.GetLatestLedgerSequence()
+	if err != nil {
+		return errors.Wrap(err, "Error getting the latest ledger sequence")
+	}
+
+	fmt.Printf("Initializing state for ledger=%d\n", s.currentLedger)
```
Comment (tamirms) [inline exp/ingest/live_session.go]: when is it appropriate to use the log package vs `fmt.Printf` ?

File: exp/ingest/live_session.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+	err := s.validate()
+	if err != nil {
+		return errors.Wrap(err, "Validation error")
+	}
+
+	s.ensureRunOnce()
+
+	historyAdapter := adapters.MakeHistoryArchiveAdapter(s.Archive)
+
+	s.currentLedger, err = historyAdapter.GetLatestLedgerSequence()
```
Comment (tamirms) [inline exp/ingest/live_session.go]: can this be a local variable instead of a member variable of `LiveSession`?

File: exp/ingest/live_session.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	for {
+		ledgerReader, err := ledgerAdapter.GetLedger(s.currentLedger)
+		if err != nil {
+			if err == io.ErrNotFound {
+				fmt.Println("Waiting for the next ledger close...")
+				time.Sleep(time.Second)
+				continue
+			}
+
+			return errors.Wrap(err, "Error getting ledger")
```
Comment (tamirms) [inline exp/ingest/live_session.go]: it seems like once you run into an error the session is kind of destroyed. there's no way to retry from the last successfully processed ledger, right? I think there might be cases where we encounter transient errors and it would be nice to support different retry strategies.

I don't know if this issue should be addressed in this PR but just wanted to leave a note

File: exp/ingest/live_session.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+			return nil
+		}
+
+		s.currentLedger++
+	}
+
+	return nil
+}
+
+func (s *LiveSession) SetStatePipeline(p *pipeline.StatePipeline) {
```
Comment (tamirms) [inline exp/ingest/live_session.go]: it seems like the state pipeline and ledger pipeline should be constructor / struct parameters since we cannot change the pipelines once the session has started running

File: exp/ingest/main.go [hunk 8]
```diff
@@ -1,9 +1,53 @@
...
+
+// LiveSession initializes the ledger state using `Archive` and `StatePipeline`,
+// then starts processing ledger data using `ledgerbackend`.
+type LiveSession struct {
+	standardSession
+
+	Archive       historyarchive.ArchiveInterface
+	LedgerBackend ledgerbackend.LedgerBackend
+
+	// mutex is used to make sure queries across many stores are persistent
```
Comment (tamirms) [inline exp/ingest/main.go]: I think these 2 lines can be deleted

File: exp/ingest/main.go [hunk 9]
```diff
@@ -1,9 +1,53 @@
...
+type SingleLedgerSession struct {
+	standardSession
+
+	Archive        *historyarchive.Archive
+	LedgerSequence uint32
+
+	statePipeline *pipeline.StatePipeline
 }
 
 type Session interface {
```
Comment (tamirms) [inline exp/ingest/main.go]: is this interface necessary? I don't think I've seen it used anywhere

File: exp/ingest/single_ledger_session.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+	err = s.processState(historyAdapter, sequence)
+	if err != nil {
+		return errors.Wrap(err, "processState errored")
+	}
+
+	return nil
+}
+
+func (s *SingleLedgerSession) SetStatePipeline(p *pipeline.StatePipeline) {
```
Comment (tamirms) [inline exp/ingest/single_ledger_session.go]: I think `statePipeline` should be a struct / constructor parameter for the same reasons I mentioned in the LiveSession comment

File: exp/ingest/processors/doc.go [hunk 11]
```diff
@@ -0,0 +1,2 @@
+// processors package contain the most commonly used processors.
```
Comment (ire-and-curses) [inline exp/ingest/processors/doc.go]:
```suggestion
// processors package contains the most commonly used processors.
```

File: exp/ingest/processors/csv_printer.go [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+func (p *CSVPrinter) fileHandle() (*os.File, error) {
+	if p.Filename == "" {
+		return os.Stdout, nil
+	}
+
+	return os.Create(p.Filename)
+}
+
+func (p *CSVPrinter) ProcessState(ctx context.Context, store *pipeline.Store, r io.StateReadCloser, w io.StateWriteCloser) error {
```
Comment (ire-and-curses) [inline exp/ingest/processors/csv_printer.go]: Fine for now but we should add docstrings before general release

File: exp/ingest/io/ledger_read_closer.go (hunk 1 above)
Comment (bartekn) [reply exp/ingest/io/ledger_read_closer.go]: Added it here if you're interested: https://github.com/dominikh/go-tools/issues/529

File: exp/ingest/io/ledger_read_closer.go (hunk 2 above)
Comment (bartekn) [reply exp/ingest/io/ledger_read_closer.go]: Yes, we'll decide on #1433 next week and change all the instances in another PR.

File: exp/ingest/live_session.go (hunk 3 above)
Comment (bartekn) [reply exp/ingest/live_session.go]: It's to [ensure interface implementation](https://golang.org/doc/effective_go.html#blank_implements). It ensures that `LiveSession` implements all methods of `Session` interface. Helpful if new methods are added to the interface (compilation will fail).

File: exp/ingest/live_session.go (hunk 4 above)
Comment (bartekn) [reply exp/ingest/live_session.go]: You're right this should be removed. Created this issue: https://github.com/stellar/go/issues/1459

File: exp/ingest/live_session.go (hunk 6 above)
Comment (bartekn) [reply exp/ingest/live_session.go]: Created an issue here: #1473.

File: exp/ingest/main.go (hunk 9 above)
Comment (bartekn) [reply exp/ingest/main.go]: It's not used internally but packages that are using `ingest` can make use of it.

**Comment by bartekn** [comment]
@ire-and-curses @tamirms you can take a look again. Added a few improvements and `horizon-demo` is fully functional now.

File: exp/ingest/io/ledger_transaction.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+
+// GetChanges returns a developer friendly representation of LedgerEntryChanges.
+// Currently it results operations related LedgerEntryChanges only.
+// TODO this should include TransactionMetaV1.txChanges and fee changes too!
+func (t *LedgerTransaction) GetChanges() []Change {
+	changes := []Change{}
+
+	for _, operationMeta := range t.Meta.OperationsMeta() {
+		ledgerEntryChanges := operationMeta.Changes
+		for i := 0; i < len(ledgerEntryChanges); i++ {
```
Comment (tamirms) [inline exp/ingest/io/ledger_transaction.go]: could you use a range for loop here?

 `for i, entryChange := range ledgerEntryChanges {`

File: exp/ingest/main.go [hunk 14]
```diff
@@ -1,9 +1,76 @@
...
+// can be found in this package.
 type Session interface {
+	// Run start the session and works as long as the session is active. If
+	// you want to resume a session use Resume().
 	Run() error
+	// Resume resumes the session at ledger with a given sequence. It's up to
+	// Session user to determine what was the last ledger processed by a
+	// Session as it's stateless (or if Run() should be called first).
+	Resume(ledgerSequence uint32) error
+	GetLatestProcessedLedger() uint32
```
Comment (tamirms) [inline exp/ingest/main.go]: missing doc string here

File: exp/tools/horizon-demo/pipelines.go [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+		// Acquire write lock
+		session.UpdateLock()
+		defer session.UpdateUnlock()
+
+		// Run commits simultaneously
+		var wg sync.WaitGroup
+		wg.Add(2)
+
+		go func() {
+			err = orderBookGraph.Apply()
```
Comment (tamirms) [inline exp/tools/horizon-demo/pipelines.go]: why apply the updates here instead of applying the updates at the end of `ProcessState()` and `ProcessLedger()` ?

File: exp/ingest/io/ledger_transaction.go [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+			switch entryChange.Type {
+			case xdr.LedgerEntryChangeTypeLedgerEntryCreated:
+				created := entryChange.MustCreated()
+				changes = append(changes, Change{
+					Type: created.Data.Type,
+					Pre:  nil,
+					Post: &created.Data,
+				})
+			case xdr.LedgerEntryChangeTypeLedgerEntryUpdated:
+				state := ledgerEntryChanges[i-1].MustState()
```
Comment (tamirms) [inline exp/ingest/io/ledger_transaction.go]: I'm not familiar with the operations meta data format. Could it be possible that `ledgerEntryChanges[0]` has type `xdr.LedgerEntryChangeTypeLedgerEntryUpdated` or `xdr.LedgerEntryChangeTypeLedgerEntryRemoved` ? if that is a possibility then `ledgerEntryChanges[i-1]` would crash

File: exp/ingest/main.go (hunk 9 above)
Comment (tamirms) [reply exp/ingest/main.go]: The Golang maintainers have the following opinion on defining interfaces:

"Do not define interfaces before they are used: without a realistic example of usage, it is too difficult to see whether an interface is even necessary, let alone what methods it ought to contain."

From https://github.com/golang/go/wiki/CodeReviewComments#interfaces

Also, the fact that `SingleLedgerSession` does not support `Resume(ledgerSequence uint32) error` is another reason which makes me think the interface is not necessary.

That being said, I will defer to your judgement on whether it makes sense to keep the interface

File: exp/tools/horizon-demo/pipelines.go (hunk 15 above)
Comment (bartekn) [reply exp/tools/horizon-demo/pipelines.go]: It's to make sure that state (db state and graph state) is consistent (at the same ledger). If we were applying and committing in the pipeline it's possible that for a short time data in both stores (db and memory) would represent state of 2 different ledgers.

File: exp/ingest/io/ledger_transaction.go (hunk 16 above)
Comment (bartekn) [reply exp/ingest/io/ledger_transaction.go]: It has specific format and I checked it in [stellar-core](https://github.com/stellar/stellar-core/blob/master/src/ledger/LedgerTxn.cpp#L582). The algorithm is:
1. If there is an existing entry:
   1. Insert `STATE`.
   2. Insert `UPDATED` or `REMOVED`.
2. Otherwise insert `CREATED`.

Comment (tamirms) [reply exp/ingest/io/ledger_transaction.go]: does that mean `LedgerEntryChanges` is always constructed such that every `UPDATED` and `REMOVED` entry is always preceded with a  `STATE` which represents the state prior to the update / removal?

File: exp/tools/horizon-demo/orderbook_processor.go [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+		default:
+			continue
+		}
+	}
+
+	return nil
+}
+
+func (p *OrderbookProcessor) IsConcurrent() bool {
+	return true
```
Comment (tamirms) [inline exp/tools/horizon-demo/orderbook_processor.go]: I don't think this processor should be concurrent because it is important that the offers be added to the graph in order. I think if the processor is concurrent then that means ledger and state entries could be processed out of order

File: exp/tools/horizon-demo/pipelines.go (hunk 15 above)
Comment (tamirms) [reply exp/tools/horizon-demo/pipelines.go]: I think a lot of this logic should be encapsulated in `DatabaseProcessor` and  `OrderbookProcessor`. `DatabaseProcessor` could define `Begin()` and `Commit()` functions which are then called in the pre / post processing hooks. Similarly, `OrderbookProcessor` could define an `Appy()` function which would be called in the post processing hook.

File: exp/support/pipeline/main.go [hunk 18]
```diff
@@ -39,14 +41,32 @@ type multiWriteCloser struct {
...
 	cancelledMutex sync.Mutex
 	cancelled      bool
+	cancelFunc     context.CancelFunc
 }
 
+// PipelineInterface is an interface that defines common pipeline methods
+// in structs that embed Pipeline.
+type PipelineInterface interface {
+	SetRoot(rootProcessor *PipelineNode)
+	AddPreProcessingHook(hook func(context.Context) error)
```
Comment (tamirms) [inline exp/support/pipeline/main.go]: could `Reset()` be implemented as a pre-processing hook? how are those two operations different?

File: exp/ingest/io/ledger_transaction.go (hunk 16 above)
Comment (bartekn) [reply exp/ingest/io/ledger_transaction.go]: Yes, it results from the algorithm above. However, I will confirm it with the core team next week.

File: exp/tools/horizon-demo/orderbook_processor.go (hunk 17 above)
Comment (bartekn) [reply exp/tools/horizon-demo/orderbook_processor.go]: Is it really important that offers must be added in order? For initial processing it doesn't matter. For ledger/transactions processing I think it also doesn't matter because graph is locked for reading during updates. Can you elaborate?

File: exp/support/pipeline/main.go (hunk 18 above)
Comment (bartekn) [reply exp/support/pipeline/main.go]: You mean internal `reset()` function is always added to `preProcessingHooks` when pipeline is created? I think we can do it, but when we decide on #1433. If someone creates new pipeline not via constructor, then the reset hook won't be added.

File: exp/tools/horizon-demo/orderbook_processor.go (hunk 17 above)
Comment (tamirms) [reply exp/tools/horizon-demo/orderbook_processor.go]: what if someone creates an offer and then later decides to remove the offer? or if someone creates an offer and then the offer is updated because the offer is partially consumed? it's important that those events are processed in order because if you try to remove an offer which doesn't exist the code will panic

Comment (bartekn) [reply exp/tools/horizon-demo/orderbook_processor.go]: OK, you're totally right!

File: exp/tools/horizon-demo/pipelines.go (hunk 15 above)
Comment (bartekn) [reply exp/tools/horizon-demo/pipelines.go]: In general I agree but I think the problem is that we have two `DatabaseProcessor` instances: one for inserting transactions, the other for updating signers. Then which one we use to call `Begin()` and `Commit()`. We can call `Begin()` only once, it would be confusing if it's called on a single processor. That's why I decided to call it on `Database` directly.

**Comment by bartekn** [comment]
Added some updates connected to feedback. PTAL!

**Comment by tamirms** [comment]
@bartekn it looks good! the only remaining comments I have is that the mutex in `orderBookBatchedUpdates` is no longer necessary because the orderbook processor is not concurrent. Also, I think it would be better if updates could only be applied in one way. My suggestion is to either

* get rid of the `AddOffer`, `RemoveOffer`, and `Apply` methods on `OrderBookGraph`. Instead, you can create the batch and apply it in the pipeline hooks similar to what you do with the database transaction for the `DatabaseProcessor` instances

Or

* get rid of the `BatchedUpdates` interface entirely so that you cannot create multiple batches. The only way to apply updates would be to call the `AddOffer`, `RemoveOffer`, and `Apply` methods on `OrderBookGraph`.

File: exp/orderbook/batch.go [hunk 19]
```diff
@@ -61,15 +39,8 @@ func (tx *orderBookBatchedUpdates) AddOffer(offer xdr.OfferEntry) BatchedUpdates
...
-func (tx *orderBookBatchedUpdates) RemoveOffer(offerID xdr.Int64) BatchedUpdates {
-	tx.mutex.Lock()
-	defer tx.mutex.Unlock()
-
-	if tx.committed {
-		panic(errBatchAlreadyApplied)
-	}
-
+// removeOffer will queue an operation to remove the given offer from the order book
+func (tx *orderBookBatchedUpdates) RemoveOffer(offerID xdr.Int64) *orderBookBatchedUpdates {
```
Comment (tamirms) [inline exp/orderbook/batch.go]: should be `removeOffer`
